src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -173,15 +173,8 @@ await foreach (var producer in _channel.Reader.ReadAllAsync())
                     stream.ResponseTrailers.SetReadOnly();
                     stream.DecrementActiveClientStreamCount();
 
-                    if (buffer.Length > 0)
-                    {
-                        // It is faster to write data and trailers together. Locking once reduces lock contention.
-                        flushResult = await WriteDataAndTrailersAsync(stream, buffer, producer.WriteHeaders, stream.ResponseTrailers);
-                    }
-                    else
-                    {
-                        flushResult = await WriteResponseTrailersAsync(stream, producer.WriteHeaders, stream.ResponseTrailers);
-                    }
+                    // It is faster to write data and trailers together. Locking once reduces lock contention.
+                    flushResult = await WriteDataAndTrailersAsync(stream, buffer, producer.WriteHeaders, stream.ResponseTrailers);
```

Comment:
I combined these to reduce grpc requests fighting over the lock. Theoretically, they don't need to be merged anymore if only one thread is taking the write lock (and the occasional window updates).

I don't think anything needs to be done now, should wait for perf data first, but maybe they should be separate methods.

Reply:
I don't think so. It's actually nicer have them all combined into a single method.